File: SoftUni - C# OOP/Encapsulation/Encapsulation - Exercise/04. Pizza Calories/Dough.cs
using System;

namespace PizzaCalories
{
    class Dough
    {
        private const int MIN_WEIGHT = 1;
        private const int MAX_WEIGHT = 200;
        private const int BASE_CALORIES = 2;

        private string flourType;
        private string bakingTechnique;
        private double weight;
        public Dough(string flourType, string bakingTechnique, double weight)
        {
            this.FlourType = flourType;
            this.BakingTechnique = bakingTechnique;
            this.Weight = weight;
        }
        public string FlourType
        {
            get
            {
                return this.flourType;
            }
            set
            {
                string type = value.ToLower();
                if (type != "white" && type != "wholegrain")
                {
                    throw new ArgumentException("Invalid type of dough.");
                }
                this.flourType = value;
            }
        }

        public string BakingTechnique
        {
            get
            {
                return this.bakingTechnique;
            }
            set
            {
                string foodProducts = value.ToLower();
                if (foodProducts != "crispy" && foodProducts != "chewy" && foodProducts != "homemade")
                {
                    throw new ArgumentException("Invalid type of dough.");
                }
                this.bakingTechnique = value;
            }
        }

        public double Weight
        {
            get
            {
                return this.weight;
            }
            set
            {
                if (value < MIN_WEIGHT || value > MAX_WEIGHT)
                {
                    throw new ArgumentException($"Dough weight should be in the range [{MIN_WEIGHT}..{MAX_WEIGHT}].");
                }
                this.weight = value;
            }
        }
        public string Calories()
        {
            double value = 0;

            switch (this.flourType.ToLower())
            {
                case "white": 
                value = 1.5; 
                break;

                case "wholegrain": 
                value = 1.0; 
                break;
            }
            switch (this.bakingTechnique.ToLower())
            {
                case "crispy": 
                value *= 0.9; 
                break;

                case "chewy": 
                value *= 1.1; 
                break;

                case "homemade": 
                value *= 1.0; 
                break;
            }
            double calories = (BASE_CALORIES * this.weight) * value;
            return calories.ToString("F2");
        }   
    }
}

File: SoftUni - Java OOP/Encapsulation/Encapsulation Exercises/_04_Pizza_Calories/Dough.java
package _04_Pizza_Calories;

public class Dough extends CaloriesCalculator {
    private String flourType;
    private String bakingTechnique;
    private double weight;

    public Dough(String flourType, String bakingTechnique, double weight) {
        setFlourType(flourType);
        setBakingTechnique(bakingTechnique);
        setWeight(weight);
    }

    private void setFlourType(String flourType) {
        if (!flourType.equals("White") && !flourType.equals("Wholegrain")) {
            throw new IllegalArgumentException("Invalid type of dough.");
        }
        this.flourType = flourType;
    }

    private void setBakingTechnique(String bakingTechnique) {
        if (!bakingTechnique.equals("Crispy") && !bakingTechnique.equals("Chewy") && !bakingTechnique.equals("Homemade")) {
            throw new IllegalArgumentException("Invalid type of dough.");
        }
        this.bakingTechnique = bakingTechnique;
    }

    private void setWeight(double weight) {
        if (weight < 1 || weight > 200) {
            throw new IllegalArgumentException("Dough weight should be in the range [1..200].");
        }
        this.weight = weight;
    }

    @Override
    public double calculateCalories() {
        double baseCalories = 2 * weight;
        double flourModifier = flourType.equals("White") ? 1.5 : 1.0;
        double techniqueModifier = switch (bakingTechnique) {
            case "Crispy" -> 0.9;
            case "Chewy" -> 1.1;
            case "Homemade" -> 1.0;
            default -> 0;
        };
        return baseCalories * flourModifier * techniqueModifier;
    }
}
